[DOTNET] Create Utility class for Regular Expression

GiulianoDolceTalianoSantoro opened the pull request:
Solves [ISSUE#1427](https://github.com/microsoft/Recognizers-Text/issues/1427)

### Description
This PR will centralize all Regular expressions related methods in a class called `RegExpUtility` in .NET. The idea here is to increase parity to match how it is already implemented in Java, JavaScript and Python. 

### Changes made

- We added four more methods from `RegexExtensions.cs`:
   - `IsExactMatch( )`
   - `MatchExact( )`
   - `MatchEnd( )`
   - `MatchBegin( )`
 
- We changed `MatchEvaluator( )` to `ReplaceMatchGroup( )` because that name was more descriptive

- We moved the `ConditionalMatch` class from `Microsoft.Recognizers.Text.DateTime` to `Microsoft.Recognizers.Text.Utilities`, so the methods we moved before do not break the existing references.
 
- We imported `Microsoft.Recognizers.Text.Utilities` everywhere the code had errors because of the classes and methods migration.

### Testing
Find below a screenshot of the running tests

![image](https://user-images.githubusercontent.com/54330145/64641066-c18f3100-d3e1-11e9-87f2-e5fe172c68a2.png)

GiulianoDolceTalianoSantoro committed:
Create RegExpUtility class with methods

Created
* GetMatches( )
* MatchEnd( )
* GetFirstMatchIndex( )
* Split( )
* SanitizeGroups( )

Removed
* GetMatchesSimple( ): In JavaScript was only used in GetMatches method, but this method already existed in .NET, and it did not need GetMatchesSimple, so I removed it.
* GetSafeRegex( ): In JavaScript It was only used in QueryProcessor class, which, in .NET, didn't need it.
* IsMatch( ): in .NET is a Regex function so I didn't need to create it
* GetNextRegex( ): It wasn't use anywhere in the code
* GetClosePos( ): It was only used in GetNextRegex method, which I removed

GiulianoDolceTalianoSantoro committed:
Add all methods into RegExpUtility class and migrate ConditionalMatch class, and RegexExtensios.cs methods too.

We added four more methods from RegexExtensions.cs:
* IsExactMatch( )
* MatchExact( )
* Match End( )
* MatchBegin( )

And removed one:
* GetFirstMatchIndex( ): It had not use anywhere in dotnet

We change MatchEvaluator( ) to ReplaceMatchGroup( ) because that name was more descriptive

Also, we moved ConditionalMatch class from Microsoft.Recognizers.Text.DateTime to Microsoft.Recognizers.Text.Utilites, so the methods we moved before do not break the existing references.

Finally, we imported Microsoft.Recognizers.Text.Utilites everywhere the code had errors because of the classes and methods migration.

VictorGrycuk commented on .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs:
```diff
@@ -359,11 +359,11 @@ protected ParseResult FracLikeNumberParse(ExtractResult extResult)
                 var smallValue = char.IsDigit(numerator[0]) ?
                     GetDigitalValue(numerator, 1) :
-                    GetIntValue(GetMatches(numerator));
+                    GetIntValue(Utilities.RegExpUtility.GetMatches(this.TextNumberRegex, numerator)); // ACAACA
```
Is the comment `// ACAACA` used for something?

GiulianoDolceTalianoSantoro replied:
It was only to remember to return to that line when looking for "ACAACA" in the code, but I forgot to delete it. I already solved it, thanks for the feedback.

VictorGrycuk commented on .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs:
```diff
@@ -696,21 +696,21 @@ private bool SkipNonDecimalSeparator(char ch, int distance)
-
-            return matchStrs;
-        }
+        // private List<string> GetMatches(string input)
```
This whole method was commented, shouldn't it be removed instead?

GiulianoDolceTalianoSantoro replied:
It should, I just forgot about it, but I removed it already, thank you!

Maihav commented on .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs:
```diff
@@ -696,21 +696,21 @@ private bool SkipNonDecimalSeparator(char ch, int distance)
+        //    }
+
+        // return matchStrs;
+        // }
```
Why is this method commented?

GiulianoDolceTalianoSantoro replied:
I didn't suppose to let that commented, but I forgot. I already removed it, thanks for the feedback!

ceciliaavila commented on .NET/Microsoft.Recognizers.Text.Number/Parsers/BaseNumberParser.cs:
```diff
@@ -696,21 +696,21 @@ private bool SkipNonDecimalSeparator(char ch, int distance)
+        //    }
+
+        // return matchStrs;
+        // }
```
Please remove commented code.

GiulianoDolceTalianoSantoro replied:
For some reason I actually see those files in my branch, I don't really know why they appear as deleted in the first commit (matcher.py & abstract_matcher.py). I'll let that for now as it is, thank you for pointing that and for the feedback!

![image](https://user-images.githubusercontent.com/54330145/64701632-82172200-d47f-11e9-8154-18654ccaf9bc.png)

GiulianoDolceTalianoSantoro replied:
The problem was solved by deleting both Matcher and matcher folders (matcher was not supposed to be created, but for some reason it created itself as a Matcher clone, with the files that were supposedly deleted, an issue in Microsoft's Recognizers-Text repository) and adding Matcher again with the deleted Python files (as they were locally in my Matcher folder). We're going to make a PR to fix this in MS repository. Thanks for the feedback and for making us realize that.

MS & SW repository: 

![image](https://user-images.githubusercontent.com/54330145/64791850-b574b200-d54e-11e9-9486-9e73ca4a3fb8.png)

feature/create-regex-utility-class branch (PR #304)

![image](https://user-images.githubusercontent.com/54330145/64791947-e5bc5080-d54e-11e9-99b5-c16d636d1c22.png)

ceciliaavila reviewed (changes requested):
Just some minor feedback and a question. Why are you deleting .py files in here?

Konigstiger reviewed (approved):
LGTM, with the already proposed modifications of 
- innecesary comment.
- remove code commented method.

GiulianoDolceTalianoSantoro committed:
Remove commented code

I removed the unnecesary comments  I made in BaseNumberParser.cs

GiulianoDolceTalianoSantoro committed:
Resolving erased python files

As some files exist locally in my branch but they're shown as deleted in the repository, I'm deleting them to see if there are any changes.

GiulianoDolceTalianoSantoro committed:
Revert "Resolving erased python files"

This reverts commit 7108bfd100911a7f7a45147d440672321f259b7e.

GiulianoDolceTalianoSantoro committed:
Changing folder name

Aliandi committed:
Delete abstract_matcher.py

Aliandi committed:
Delete matcher.py

GiulianoDolceTalianoSantoro committed:
Adding erased python files

GiulianoDolceTalianoSantoro committed:
Adding erased Python files

GiulianoDolceTalianoSantoro committed:
Deleting Matcher folder

GiulianoDolceTalianoSantoro committed:
Adding Matcher folder

Aliandi committed:
Merge branch 'feature/create-regex-utility-class' of https://github.com/southworks/Recognizers-Text into feature/create-regex-utility-class

Aliandi committed:
rename

Aliandi committed:
rename to correct case

Aliandi committed:
fix python build error

Aliandi committed:
fix imports from tests

Aliandi committed:
fix remaining imports

Aliandi committed:
re-add deleted matcher file

Aliandi committed:
readd abstract matcher
